dbcore/sm-dir-it.h

```diff
@@ -0,0 +1,61 @@
+/* 
+ * OID Dir Iterator, used for Index query
+ */
+#pragma once
+#include "sm-config.h"
+#include "../txn.h"
+namespace ermia {
+
+struct DirIterator {
+  /* For debug use, store the oid_dir ptr */
+  OID *dirp;
+  TableDescriptor *td;
+  transaction *t;
+  int idx = 0;
+  DirIterator(transaction *t, TableDescriptor *td) : td(td), t(t) {
+
+  };
+
+  ~DirIterator() {
+      
+  }
+
+  const ermia::varstr next(bool &eof) {
+redo:
+    ermia::varstr tmpval; 
+    auto o = oidmgr->dir_get_index(td->GetTupleArray(), dirp, idx + 1, eof);
+    if (eof) {
+        return tmpval;
+    }
+    dbtuple *tuple = nullptr;
+    tuple = oidmgr->oid_get_version(td->GetTupleArray(), o, t->xc);
+    bool found = true;
+    if (!tuple) {
+        DLOG(WARNING) << "(SKIPPED) Some tuple is empty: OID = " << std::hex << o;
+        found = false;
+    }
+    if (found) {
+        bool ret = t->DoTupleRead(tuple, &tmpval)._val;
```

Comment:
`DoTupleRead` returns rc_t, not boolean. If the return value indicates an abort, the calling transaction should abort. (not sure why a retry is needed here?)

Reply:
Sorry for the inconvenience, the changes are checked in the new branch `reopen`, basically change the ret type to rc_t and also ALWAYS_ASSERT ret is RC_TRUE